### PR TITLE
Fix: Expected order of jsdoc tags (fixes #9412)

### LINKED ISSUE #9412: Invalid return when jsdoc @abstract is after @return(s)
**Tell us about your environment**

* **ESLint Version:**
v4.7.2

* **Node Version:**
v7.1.0

* **npm Version:**
v3.10.9

**What parser (default, Babel-ESLint, etc.) are you using?**
Default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js

module.exports = {
  'extends': 'google',
  'parserOptions': {
    'ecmaVersion': 6
  },
  'rules': {
    'prefer-rest-params': 'off',
    'prefer-spread': 'off',
    'eqeqeq': 'error',
    // Override the google eslint config to prefer @returns
    'valid-jsdoc': [
      'error',
      {
        'requireParamDescription': false,
        'requireReturnDescription': false,
        'requireReturn': false,
        'prefer': {
          'return': 'returns'
        }
      }
    ]
  }
};


```

</details>

**What did you do? Please include the actual source code causing the issue.**

```js

  /**
   * @returns {string}
   * @abstract
   */
 someMethod() {}

```

**What did you expect to happen?**

It [doesn't seem like](https://stackoverflow.com/questions/24306839/is-there-an-official-order-for-jsdoc-tags-in-documentation) there's an expected order for most JSDoc tags? So I would expect order to not matter between the `@abstract` and `@returns` tags. But in the implementation of this [feature request](https://github.com/eslint/eslint/issues/5941) there's an implicit expectation that `@abstract` is *before* `@returns`.

**What actually happened? Please include the actual, raw output from ESLint.**

```
error  Unexpected @returns tag; function has no return statement  valid-jsdoc
```

### PR DESCRIPTION
**Pull request by Orlandster1998**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#9412 

**What changes did you make? (Give an overview)**
I've added tests to make sure that the jsdoc tags can be used in any order. So you will find each possible combination of jsdoc tags, based on these [tags](https://github.com/SimonSchick/eslint/blob/399a1e226092bce2aedd910a539b43da6e5139e6/lib/rules/valid-jsdoc.js#L215).

I've also removed the logic from the switch cases to fix the issue. Now the switch cases will just be used to set the status flags. (e.g. hasParams...)

I think it's definitely better to separate the logic from the cases, especially when it comes to different kind of dependencies in between this status flags. (that's why the bug occurred)

**Is there anything you'd like reviewers to focus on?**

### CODE REVIEW

File: lib/rules/valid-jsdoc.js [hunk 1]
```diff
@@ -226,8 +226,11 @@ module.exports = {
         function checkJSDoc(node) {
             const jsdocNode = sourceCode.getJSDocComment(node),
                 functionData = fns.pop(),
-                params = Object.create(null);
+                params = Object.create(null),
+                hasParamsTags = [];
             let hasReturns = false,
+                hasReturnsTag,
+                hasParams = false,
```
Comment (not-an-aardvark) [inline lib/rules/valid-jsdoc.js]: I think the `has` prefix here is intended to refer to boolean conditions (e.g. the `hasReturns` variable describes the boolean condition of whether the JSDoc comment has any returns). You've created some other variables prefixed with `has`, but they don't seem to be describing boolean conditions, which was confusing to me as I was reading through the code. I would recommend removing the `has` prefix for those variables (e.g. renaming `hasReturnsTag` to `returnsTag`).

File: lib/rules/valid-jsdoc.js [hunk 2]
```diff
@@ -333,6 +307,42 @@ module.exports = {
                     }
                 });
 
+                if (hasParams) {
```
Comment (not-an-aardvark) [inline lib/rules/valid-jsdoc.js]: I think the `hasParams` variable is unnecessary. If there are no `param` tags, then the `hasParamsTags` array will be empty, so the `forEach` loop will just do nothing anyway.

**Comment by Orlandster1998** [comment]
@not-an-aardvark Thanks for the review. You're absolutely right! I've made a commit with the improvements.

**Comment by kaicataldo** [comment]
LGTM. Thanks for contributing to ESLint!

**Pull request merged**